---
title: "Extract environmental data from Bio-Oracle"
date: '2020-11-01'
diagram: yes
math: yes
slug: r-extract-environmental-data
summary: This will be a summary.
tags:
- R
- Raster
- Marine
categories: R
---



<div id="contents"/>
<div id="contents" class="section level2">
<h2>Contents</h2>
<ol style="list-style-type: decimal">
<li><a href="#1">Why bother with R?</a></li>
<li><a href="#2">Import genetic data</a></li>
<li><a href="#3">Filtering</a></li>
<li><a href="#4">Summary statistics</a></li>
<li><a href="#5"><em>F</em><sub>ST</sub>, PCA &amp; DAPC</a></li>
<li><a href="#6">Extras</a>
<ul>
<li><a href="#vcf">Import VCF file</a></li>
<li><a href="#outlier">Conduct outlier tests using OutFLANK</a></li>
<li><a href="#export">Export biallelic SNP genind object</a></li>
<li><a href="#resources">Further resources and R session info</a></li>
</ul></li>
</ol>
<div id="1"/>
<div id="why-bother-with-r" class="section level3">
<h3>1. Why bother with R?</h3>
<p>There are so many programs and software out there for analysing population genetic data and generating summary statistics. At first I was quite overwhelmed and unsure which path to take. Then I started learning and using R and I’ve not looked back since. Aside from the fact that R was the first programming language I learnt, there are several reasons why I like to use R for popgen analysis:</p>
<ul>
<li>A wealth of online help resources and tutorials</li>
<li>Analyses easily replicated on new data sets<br />
</li>
<li>Many options for creating publication quality figures and visualisations<br />
</li>
<li>Code can be uploaded to online repositories for other people to reproduce your analysis<br />
</li>
<li>Cross platform compatibility</li>
<li>Free!</li>
</ul>
<p>Nowadays, popgen R has dozens of packages that often allow you to do similar things but different packages can have their own formatting requirements and R objects. I recommend choosing one type of R object to conduct all your analyses with because converting between R objects can be difficult and frustrating (I work with <code>genind</code> objects from the <a href="https://github.com/thibautjombart/adegenet">adegenet</a> package).</p>
<p>In this post, I cover some ‘bread-and-butter’ analyses for typical popgen data sets and highlight some of the R packages and functions I use to analyse such data using example data sets from published studies.</p>
<p><strong>Assumptions</strong><br />
This post assumes that you have installed R and RStudio and that you have some skills in R coding and functionality. To follow along, I recommend that you download the example data sets to a directory of your choice, create a new R script in the same directory and then set your working directory to the location of these files. To set your working directory in RStudio, for example, click Session &gt; Set Working Directory &gt; To Source File Location.</p>
<p><strong>Download example data sets</strong><br />
<a href="/files/Lobster_SNP_Genotypes.csv">1. European lobster SNP genotypes in data.frame format</a><br />
<a href="/files/Pinkseafan_13MicrosatLoci.gen">2. Pink sea fan microsatellite genotypes in genepop format</a></p>
<p><strong>References</strong><br />
Jenkins TL, Ellis CD, Triantafyllidis A, Stevens JR (2019). <a href="https://doi.org/https://doi.org/10.1111/eva.12849">Single nucleotide polymorphisms reveal a genetic cline across the north‐east Atlantic and enable powerful population assignment in the European lobster</a>. <em>Evolutionary Applications</em> 12, 1881–1899.<br />
Holland LP, Jenkins TL, Stevens JR (2017). <a href="https://doi.org/https://doi.org/10.1038/hdy.2017.14">Contrasting patterns of population structure and gene flow facilitate exploration of connectivity in two widely distributed temperate octocorals</a>. <em>Heredity</em> 119, 35–48.</p>
<p><a href="#contents">Back to Contents</a></p>
<div id="2"/>
</div>
<div id="import-genetic-data" class="section level3">
<h3>2. Import genetic data</h3>
</div>
</div>
