---
title: "SUB"
date: 2020-09-25
math: yes
diagram: yes
image:
# Featured image placement options:
#   1 = Full column width
#   2 = Out-set
#   3 = Screen-width
  caption: "NA"
  placement: 1
summary: "R provides a wealth of packages and resources for analysing population genetic data. This post aims to help scientists at any research level get started with analysis of genetic data using R and its packages."
tags: ["Population genetics", "R", "adegenet", "genind object", "SNP", "microsatellite"]
categories: ["R"]
slug: "r-popgen-getting-started"

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

<div id="contents"/>
## Contents
1. [Why bother with R?](#1)
1. [Import genetic data](#2)
1. [Filtering](#3)
1. [Summary statistics](#4)
1. [_F_~ST~, PCA & DAPC](#5)
1. [Extras](#6)
    - [Import VCF file](#vcf)
    - [Conduct outlier tests using OutFLANK](#outlier)
    - [Export biallelic SNP genind object](#export)

<div id="1"/>
### 1. Why bother with R?
There are so many programs and software out there for analysing population genetic data and generating summary statistics. At first I was quite overwhelmed and unsure which path to take. Then I started learning and using R and I have not looked back since. Aside from the fact that R was the first programming language I learnt, there are several reasons why I like to use R for popgen analysis:

* A wealth of online help resources and tutorials
* Analyses easily replicated on new data sets  
* Many options for creating publication quality figures and visualisations  
* Code can be uploaded to online repositories for other people to reproduce your   analysis  
* Cross platform compatibility
* Free!

Nowadays, popgen R has dozens of packages that often allow you to do similar things but different packages can have their own formatting requirements and R objects. I recommend choosing one type of R object to conduct all your analyses with because converting between R objects can be difficult and frustrating (I work with `genind` objects from the `adegenet` package).

In this post, I cover some 'bread-and-butter' analyses for typical popgen data sets and highlight some of the R packages and functions I use to analyse such data using example data sets from published studies.

**Assumptions**  
This post assumes that you have installed R and RStudio and that you have some skills in R coding and functionality. To follow along, I recommend that you download the example data sets to a directory of your choice, create a new R script in the same directory and then set your working directory to the location of these files. To set your working directory in RStudio, for example, click Session > Set Working Directory > To Source File Location.

**Download example data sets**  
[1. European lobster SNP genotypes in data.frame format](./Lobster_SNP_Genotypes.csv)  
[2. Pink sea fan microsatellite genotypes in genepop format](./Pinkseafan_13MicrosatLoci.gen)

**References**  
Jenkins TL, Ellis CD, Triantafyllidis A, Stevens JR (2019). [Single nucleotide polymorphisms reveal a genetic cline across the north‚Äêeast Atlantic and enable powerful population assignment in the European lobster](https://doi.org/https://doi.org/10.1111/eva.12849). _Evolutionary Applications_ 12, 1881--1899.  
Holland LP, Jenkins TL, Stevens JR (2017). [Contrasting patterns of population structure and gene flow facilitate exploration of connectivity in two widely distributed temperate octocorals](https://doi.org/https://doi.org/10.1038/hdy.2017.14). _Heredity_ 119, 35--48.

[Back to Contents](#contents)

