---
title: "Getting started with population genetics using R"
date: '2020-09-21'
diagram: yes
math: yes
slug: r-popgen-getting-started
summary: R provides a wealth of packages and resources for analysing population genetic data. This tutorial aims to help scientists at any research level get started with analysis of genetic data using R and its packages.
categories: R
weight: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

<div id="contents"/>
## Contents
1. [Why bother with R?](#1)
1. [Import genetic data](#2)
1. [Filtering](#3)
1. [Summary statistics](#4)
1. [_F_~ST~, PCA & DAPC](#5)
1. [Extras](#6)
    - [Import VCF file](#vcf)
    - [Conduct outlier tests using OutFLANK](#outlier)
    - [Export biallelic SNP genind object](#export)
    - [Further resources and R session info](#resources)

<div id="1"/>
### 1. Why bother with R?
There are so many programs and software out there for analysing population genetic data and generating summary statistics. At first I was quite overwhelmed and unsure which path to take. Then I started learning and using R and I've not looked back since. Aside from the fact that R was the first programming language I learnt, there are several reasons why I like to use R for popgen analysis:

* A wealth of online help resources and tutorials
* Analyses easily replicated on new data sets  
* Many options for creating publication quality figures and visualisations  
* Code can be uploaded to online repositories for other people to reproduce your analysis
* Cross platform compatibility
* Free!

Nowadays, popgen R has dozens of packages that often allow you to do similar things but different packages can have their own formatting requirements and R objects. I recommend choosing one type of R object to conduct all your analyses with because converting between R objects can be difficult and frustrating (I work with `genind` objects from the [adegenet](https://github.com/thibautjombart/adegenet) package).

In this post, I cover some 'bread-and-butter' analyses for typical popgen data sets and highlight some of the R packages and functions I use to analyse such data using example data sets from published studies.

**Assumptions**  
This post assumes that you have installed R and RStudio and that you have some skills in R coding and functionality. To follow along, I recommend that you download the example data sets to a directory of your choice, create a new R script in the same directory and then set your working directory to the location of these files. To set your working directory in RStudio, for example, click Session > Set Working Directory > To Source File Location.

**Download example data sets**  
[1. European lobster SNP genotypes in data.frame format](/files/Lobster_SNP_Genotypes.csv)  
[2. Pink sea fan microsatellite genotypes in genepop format](/files/Pinkseafan_13MicrosatLoci.gen)

**References**  
Jenkins TL, Ellis CD, Triantafyllidis A, Stevens JR (2019). [Single nucleotide polymorphisms reveal a genetic cline across the north‚Äêeast Atlantic and enable powerful population assignment in the European lobster](https://doi.org/https://doi.org/10.1111/eva.12849). _Evolutionary Applications_ 12, 1881--1899.  
Holland LP, Jenkins TL, Stevens JR (2017). [Contrasting patterns of population structure and gene flow facilitate exploration of connectivity in two widely distributed temperate octocorals](https://doi.org/https://doi.org/10.1038/hdy.2017.14). _Heredity_ 119, 35--48.

[Back to Contents](#contents)

<div id="2"/>
### 2. Import genetic data

#### Install and load R packages
```{r warning=FALSE, message=FALSE}
library(adegenet)
library(poppr)
library(dplyr)
library(hierfstat)
library(reshape2)
library(ggplot2)
library(RColorBrewer)
library(scales)
```

#### Import lobster SNP genotypes

Import csv file containing SNP (single nucleotide polymorphism) genotypes.
```{r}
lobster = read.csv("Lobster_SNP_Genotypes.csv")
str(lobster)
```

Convert `data.frame` from long to wide format. The wide format contains one row for each individual and one column for each locus as well as a column for the ID and site labels.
```{r}
lobster_wide = reshape(lobster, idvar = c("ID","Site"), timevar = "Locus", direction = "wide", sep = "")

# Remove "Genotype" from column names
colnames(lobster_wide) = gsub("Genotype", "", colnames(lobster_wide))
```

Subset genotypes and only keep SNP loci used in Jenkins et al. 2019.
```{r}
# Subset genotypes
snpgeno = lobster_wide[ , 3:ncol(lobster_wide)]

# Keep only SNP loci used in Jenkins et al. 2019
snps_to_remove = c("25580","32362","41521","53889","65376","8953","21197","15531","22740","28357","33066","51507","53052","53263","21880","22323","22365")
snpgeno = snpgeno[ , !colnames(snpgeno) %in% snps_to_remove]
```

Create vectors of individual and site labels.
```{r}
ind = as.character(lobster_wide$ID) # individual ID
site = as.character(lobster_wide$Site) # site ID
```

Convert `data.frame` to `genind` object. Check that the genotypes for the first five individuals and loci are as expected.
```{r}
lobster_gen = df2genind(snpgeno, ploidy = 2, ind.names = ind, pop = site, sep = "")
lobster_gen$tab[1:5, 1:10]
```

Print basic info of the `genind` object.
```{r}
lobster_gen

popNames(lobster_gen)
```


#### Import pink sea fan microsatellite genotypes

Import genepop file and convert to `genind` object. Check that the genotypes at locus Ever002 for three randomly selected individuals are as expected.
```{r}
seafan_gen = import2genind("Pinkseafan_13MicrosatLoci.gen", ncode = 3, quiet = TRUE)
set.seed(1)
tab(seafan_gen[loc = "Ever002"])[runif(3, 1, nInd(seafan_gen)), ]
```

Print basic info of the `genind` object.
```{r}
seafan_gen

popNames(seafan_gen)
```
Update the site labels so that the site code rather than the last individual label in the sample is used.
```{r}
# Use gsub to extract only letters from a vector
popNames(seafan_gen) = gsub("[^a-zA-Z]", "", popNames(seafan_gen))
popNames(seafan_gen)
```

[Back to Contents](#contents)

<div id="3"/>
### 3. Filtering

#### Missing data: loci
Calculate the percentage of complete genotypes per loci in the lobster SNP data set.

```{r}
locmiss_lobster = propTyped(lobster_gen, by = "loc")
locmiss_lobster[which(locmiss_lobster < 0.80)] # print loci with < 80% complete genotypes

# Barplot
barplot(locmiss_lobster, ylim = c(0,1), ylab = "Complete genotypes (proportion)", xlab = "Locus", las = 2, cex.names = 0.7)
```

Calculate the percentage of complete genotypes per loci in the pink sea fan microsatellite data set.
```{r}
locmiss_seafan = propTyped(seafan_gen, by = "loc")
locmiss_seafan[which(locmiss_seafan < 0.80)] # print loci with < 80% complete genotypes

# Barplot
barplot(locmiss_seafan, ylim = c(0,1), ylab = "Complete genotypes (proportion)", xlab = "Locus", las = 2, cex.names = 0.7)
```

Remove microsatellite loci with > 20% missing data.
```{r}
seafan_gen = missingno(seafan_gen, type = "loci", cutoff = 0.20)
```

#### Missing data: individuals
Calculate the percentage of complete genotypes per individual in the lobster SNP data set.
```{r}
indmiss_lobster = propTyped(lobster_gen, by = "ind")
indmiss_lobster[ which(indmiss_lobster < 0.80) ] # print individuals with < 80% complete genotypes
```

Remove individuals with > 20% missing genotypes.
```{r}
lobster_gen = missingno(lobster_gen, type = "geno", cutoff = 0.20)
```

Calculate the percentage of complete genotypes per individual in the pink sea fan microsatellite data set.
```{r}
indmiss_seafan= propTyped(seafan_gen, by = "ind")
indmiss_seafan[ which(indmiss_seafan < 0.80) ] # print individuals with < 80% complete genotypes
```

Remove individuals with > 20% missing genotypes.
```{r}
seafan_gen = missingno(seafan_gen, type = "geno", cutoff = 0.20)
```

#### Check genotypes are unique
Check all individual genotypes are unique. Duplicated genotypes can result from unintentionally sampling the same individual twice or from sampling clones.
```{r}
# Print the number of multilocus genotypes
mlg(lobster_gen)

mlg(seafan_gen)
```

Identify duplicated genotypes.
```{r}
dups_lobster = mlg.id(lobster_gen)
for (i in dups_lobster){ # for each element in the list object
  if (length(dups_lobster[i]) > 1){ # if the length is greater than 1
    print(i) # print individuals that are duplicates
  }
}

dups_seafan = mlg.id(seafan_gen)
for (i in dups_seafan){ # for each element in the list object
  if (length(dups_seafan[i]) > 1){ # if the length is greater than 1
    print(i) # print individuals that are duplicates
  }
}
```

Remove duplicated genotypes.
```{r}
# Create a vector of individuals to remove
lob_dups = c("Laz4","Eye15","Eye16","Eye01","Laz2","Eye08","Gul101","Eye25","Iom02","Hel07","Eye27","Eye05","Eye06","Eye23","Eye22","Eye11","Cro08","Tar1","Eye14","Tar3","Lyn04","Lyn15","Eye07","Eye02","Eye20")
psf_dups = c("ArmI_15")
```

```{r}
# Create a vector of individual names without the duplicates
lob_Nodups = indNames(lobster_gen)[! indNames(lobster_gen) %in% lob_dups]
psf_Nodups = indNames(seafan_gen)[! indNames(seafan_gen) %in% psf_dups]
```

```{r}
# Create a new genind object without the duplicates
lobster_gen = lobster_gen[lob_Nodups, ]
seafan_gen = seafan_gen[psf_Nodups, ]
```

```{r}
# Re-print the number of multilocus genotypes
mlg(lobster_gen)

mlg(seafan_gen)
```


#### Check loci are still polymorphic after filtering
```{r}
isPoly(lobster_gen) %>% summary

isPoly(seafan_gen) %>% summary
```

Remove loci that are not polymorphic.
```{r}
poly_loci = names(which(isPoly(seafan_gen) == TRUE))
seafan_gen = seafan_gen[loc = poly_loci]
isPoly(seafan_gen) %>% summary
```

[Back to Contents](#contents)

<div id="4"/>
### 4. Summary statistics

#### Print basic info
```{r}
lobster_gen

seafan_gen
```

#### Print the number of alleles per locus
```{r}
table(lobster_gen$loc.fac)

table(seafan_gen$loc.fac)
```

#### Print the sample size for each site
```{r}
summary(lobster_gen$pop)

summary(seafan_gen$pop)
```

#### Print the number of private alleles per site across all loci
```{r}
private_alleles(seafan_gen) %>% apply(MARGIN = 1, FUN = sum)
```

#### Print mean allelic richness per site across all loci
```{r}
allelic.richness(genind2hierfstat(seafan_gen))$Ar %>%
  apply(MARGIN = 2, FUN = mean) %>% 
  round(digits = 3)
```

#### Calculate heterozygosity per site
```{r}
# Calculate basic stats using hierfstat
basic_lobster = basic.stats(lobster_gen, diploid = TRUE)
basic_seafan = basic.stats(seafan_gen, diploid = TRUE)
```

```{r}
# Mean observed heterozygosity per site
Ho_lobster = apply(basic_lobster$Ho, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
Ho_lobster

Ho_seafan = apply(basic_seafan$Ho, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
Ho_seafan
```

```{r}
# Mean expected heterozygosity per site
He_lobster = apply(basic_lobster$Hs, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
He_lobster

He_seafan = apply(basic_seafan$Hs, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 2)
He_seafan
```

#### Visualise heterozygosity per site
```{r}
# Create a data.frame of site names, Ho and He and then convert to long format
Het_lobster_df = data.frame(Site = names(Ho_lobster), Ho = Ho_lobster, He = He_lobster) %>%
  melt(id.vars = "Site")
Het_seafan_df = data.frame(Site = names(Ho_seafan), Ho = Ho_seafan, He = He_seafan) %>%
  melt(id.vars = "Site")
```

```{r}
# Custom theme for ggplot2
custom_theme = theme(
  axis.text.x = element_text(size = 10, angle = 90, vjust = 0.5, face = "bold"),
  axis.text.y = element_text(size = 10),
  axis.title.y = element_text(size = 12),
  axis.title.x = element_blank(),
  axis.line.y = element_line(size = 0.5),
  legend.title = element_blank(),
  legend.text = element_text(size = 12),
  panel.grid = element_blank(),
  panel.background = element_blank(),
  plot.title = element_text(hjust = 0.5, size = 15, face="bold")
  )

# Italic label
hetlab.o = expression(italic("H")[o])
hetlab.e = expression(italic("H")[e])
```

```{r}
# Lobster heterozygosity barplot
ggplot(data = Het_lobster_df, aes(x = Site, y = value, fill = variable))+
  geom_bar(stat = "identity", position = position_dodge(width = 0.6), colour = "black")+
  scale_y_continuous(expand = c(0,0), limits = c(0,0.50))+
  scale_fill_manual(values = c("royalblue", "#bdbdbd"), labels = c(hetlab.o, hetlab.e))+
  ylab("Heterozygosity")+
  ggtitle("European lobster")+
  custom_theme
# Pink sea fan heterozygosity barplot
ggplot(data = Het_seafan_df, aes(x = Site, y = value, fill = variable))+
  geom_bar(stat = "identity", position = "dodge", colour = "black")+
  scale_y_continuous(expand = c(0,0), limits = c(0,0.60), breaks = c(0, 0.10, 0.20, 0.30, 0.40, 0.50, 0.60))+
  scale_fill_manual(values = c("pink", "#bdbdbd"), labels = c(hetlab.o, hetlab.e))+
  ylab("Heterozygosity")+
  ggtitle("Pink sea fan")+
  custom_theme
```

#### Inbreeding coefficient (_F_~IS~)
Calculate mean _F_~IS~ per site.
```{r}
# European lobster
apply(basic_lobster$Fis, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)

# Pink sea fan
apply(basic_seafan$Fis, MARGIN = 2, FUN = mean, na.rm = TRUE) %>%
  round(digits = 3)
```

[Back to Contents](#contents)

<div id="5"/>
### 5. _F_~ST~, PCA & DAPC

#### _F_~ST~
Compute pairwise _F_~ST~ (Weir & Cockerham 1984).
```{r}
# Subset data sets to reduce computation time
lobster_gen_sub = popsub(lobster_gen, sublist = c("Ale","Ber","Brd","Pad","Sar17","Vig"))
seafan_gen_sub = popsub(seafan_gen, sublist = c("Bla","Bov","Bre","Lun","PorI","Sko"))

# Compute pairwise Fsts
lobster_fst = genet.dist(lobster_gen_sub, method = "WC84")
lobster_fst %>% round(digits = 3)

seafan_fst = genet.dist(seafan_gen_sub, method = "WC84")
seafan_fst %>% round(digits = 3)
```

Visualise pairwise _F_~ST~ for lobster.
```{r}
# Convert dist object to data.frame
fst.matrix = as.matrix(lobster_fst)
ind = which( upper.tri(fst.matrix), arr.ind = TRUE)
fst.df = data.frame(Site1 = dimnames(fst.matrix)[[2]][ind[,2]],
                    Site2 = dimnames(fst.matrix)[[1]][ind[,1]],
                    Fst = fst.matrix[ ind ] %>% round(digits = 3))

# Convert minus values to zero
fst.df$Fst[fst.df$Fst < 0] = 0

# Print data.frame summary
fst.df %>% str

# Fst italic label
fst.label = expression(italic("F")[ST])

# Extract middle Fst value for gradient argument
mid = max(fst.df$Fst) / 2

# Plot heatmap
ggplot(data = fst.df, aes(x = Site1, y = Site2, fill = Fst))+
  geom_tile(colour = "black")+
  geom_text(aes(label = Fst), color="black", size = 3)+
  scale_fill_gradient2(low = "blue", mid = "pink", high = "red", midpoint = mid, name = fst.label, limits = c(0, max(fst.df$Fst)), breaks = c(0, 0.05, 0.10, 0.15))+
  scale_x_discrete(expand = c(0,0))+
  scale_y_discrete(expand = c(0,0), position = "right")+
  theme(axis.text = element_text(colour = "black", size = 10, face = "bold"),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position = "right",
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 10)
        )
```

Visualise pairwise _F_~ST~ for pink sea fan.
```{r}
# Convert dist object to data.frame
fst.matrix = as.matrix(seafan_fst)
ind = which( upper.tri(fst.matrix), arr.ind = TRUE)
fst.df = data.frame(Site1 = dimnames(fst.matrix)[[2]][ind[,2]],
                    Site2 = dimnames(fst.matrix)[[1]][ind[,1]],
                    Fst = fst.matrix[ ind ] %>% round(digits = 3))

# Convert minus values to zero
fst.df$Fst[fst.df$Fst < 0] = 0

# Print data.frame summary
fst.df %>% str

# Fst italic label
fst.label = expression(italic("F")[ST])

# Extract middle Fst value for gradient argument
mid = max(fst.df$Fst) / 2

# Plot heatmap
ggplot(data = fst.df, aes(x = Site1, y = Site2, fill = Fst))+
  geom_tile(colour = "black")+
  geom_text(aes(label = Fst), color="black", size = 3)+
  scale_fill_gradient2(low = "blue", mid = "pink", high = "red", midpoint = mid, name = fst.label, limits = c(0, max(fst.df$Fst)), breaks = c(0, 0.05, 0.10))+
  scale_x_discrete(expand = c(0,0))+
  scale_y_discrete(expand = c(0,0), position = "right")+
  theme(axis.text = element_text(colour = "black", size = 10, face = "bold"),
        axis.title = element_blank(),
        panel.grid = element_blank(),
        panel.background = element_blank(),
        legend.position = "right",
        legend.title = element_text(size = 14, face = "bold"),
        legend.text = element_text(size = 10)
        )
```

#### PCA

Perform a PCA (principle components analysis) on the lobster data set.
```{r}
# Replace missing data with the mean allele frequencies
x = tab(lobster_gen_sub, NA.method = "mean")

# Perform PCA
pca1 = dudi.pca(x, scannf = FALSE, scale = FALSE, nf = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent = pca1$eig/sum(pca1$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,12),
        names.arg = round(percent, 1))
```

Visualise PCA results.
```{r}
# Create a data.frame containing individual coordinates
ind_coords = as.data.frame(pca1$li)

# Rename columns of dataframe
colnames(ind_coords) = c("Axis1","Axis2","Axis3")

# Add a column containing individuals
ind_coords$Ind = indNames(lobster_gen_sub)

# Add a column with the site IDs
ind_coords$Site = lobster_gen_sub$pop

# Calculate centroid (average) position for each population
centroid = aggregate(cbind(Axis1, Axis2, Axis3) ~ Site, data = ind_coords, FUN = mean)

# Add centroid coordinates to ind_coords dataframe
ind_coords = left_join(ind_coords, centroid, by = "Site", suffix = c("",".cen"))

# Define colour palette
cols = brewer.pal(nPop(lobster_gen_sub), "Set1")

# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

# Custom theme for ggplot2
ggtheme = theme(axis.text.y = element_text(colour="black", size=12),
                axis.text.x = element_text(colour="black", size=12),
                axis.title = element_text(colour="black", size=12),
                panel.border = element_rect(colour="black", fill=NA, size=1),
                panel.background = element_blank(),
                plot.title = element_text(hjust=0.5, size=15) 
)

# Scatter plot axis 1 vs. 2
ggplot(data = ind_coords, aes(x = Axis1, y = Axis2))+
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  # spider segments
  geom_segment(aes(xend = Axis1.cen, yend = Axis2.cen, colour = Site), show.legend = FALSE)+
  # points
  geom_point(aes(fill = Site), shape = 21, size = 3, show.legend = FALSE)+
  # centroids
  geom_label(data = centroid, aes(label = Site, fill = Site), size = 4, show.legend = FALSE)+
  # colouring
  scale_fill_manual(values = cols)+
  scale_colour_manual(values = cols)+
  # custom labels
  labs(x = xlab, y = ylab)+
  ggtitle("Lobster PCA")+
  # custom theme
  ggtheme
# Export plot
# ggsave("Figure1.png", width = 12, height = 8, dpi = 600)
```

#### DAPC  
Perform a DAPC (discriminant analysis of principal components) on the seafan data set.
```{r}
# Perform cross validation to find the optimal number of PCs to retain in DAPC
set.seed(123)
x = tab(seafan_gen_sub, NA.method = "mean")
crossval = xvalDapc(x, seafan_gen_sub$pop, result = "groupMean", xval.plot = TRUE)
# Number of PCs with best stats (lower score = better)
crossval$`Root Mean Squared Error by Number of PCs of PCA`
crossval$`Number of PCs Achieving Highest Mean Success`
crossval$`Number of PCs Achieving Lowest MSE`
numPCs = as.numeric(crossval$`Number of PCs Achieving Lowest MSE`)
```

```{r}
# Run a DAPC using site IDs as priors
dapc1 = dapc(seafan_gen_sub, seafan_gen_sub$pop, n.pca = numPCs, n.da = 3)

# Analyse how much percent of genetic variance is explained by each axis
percent = dapc1$eig/sum(dapc1$eig)*100
barplot(percent, ylab = "Genetic variance explained by eigenvectors (%)", ylim = c(0,60),
        names.arg = round(percent, 1))
```

Visualise DAPC results.
```{r}
# Create a data.frame containing individual coordinates
ind_coords = as.data.frame(dapc1$ind.coord)

# Rename columns of dataframe
colnames(ind_coords) = c("Axis1","Axis2","Axis3")

# Add a column containing individuals
ind_coords$Ind = indNames(seafan_gen_sub)

# Add a column with the site IDs
ind_coords$Site = seafan_gen_sub$pop

# Calculate centroid (average) position for each population
centroid = aggregate(cbind(Axis1, Axis2, Axis3) ~ Site, data = ind_coords, FUN = mean)

# Add centroid coordinates to ind_coords dataframe
ind_coords = left_join(ind_coords, centroid, by = "Site", suffix = c("",".cen"))

# Define colour palette
cols = brewer.pal(nPop(seafan_gen_sub), "Set2")

# Custom x and y labels
xlab = paste("Axis 1 (", format(round(percent[1], 1), nsmall=1)," %)", sep="")
ylab = paste("Axis 2 (", format(round(percent[2], 1), nsmall=1)," %)", sep="")

# Scatter plot axis 1 vs. 2
ggplot(data = ind_coords, aes(x = Axis1, y = Axis2))+
  geom_hline(yintercept = 0)+
  geom_vline(xintercept = 0)+
  # spider segments
  geom_segment(aes(xend = Axis1.cen, yend = Axis2.cen, colour = Site), show.legend = FALSE)+
  # points
  geom_point(aes(fill = Site), shape = 21, size = 3, show.legend = FALSE)+
  # centroids
  geom_label(data = centroid, aes(label = Site, fill = Site), size = 4, show.legend = FALSE)+
  # colouring
  scale_fill_manual(values = cols)+
  scale_colour_manual(values = cols)+
  # custom labels
  labs(x = xlab, y = ylab)+
  ggtitle("Pink sea fan DAPC")+
  # custom theme
  ggtheme
# Export plot
# ggsave("Figure2.png", width = 12, height = 8, dpi = 600)
```

[Back to Contents](#contents)

<div id="6"/>
### 6. Extras

<div id="vcf"/>
#### Import VCF file  
To illustrate importing VCF files and conducting OutFLANK outlier selection tests in R, we will use an [American lobster SNP data set](https://doi.org/10.5061/dryad.q771r) available from the Dryad Digital Repository.
```{r warning=FALSE, message=FALSE}
# Load vcfR package
library(vcfR)

# Import only 3,000 variants to reduce computation time
american = read.vcfR("10156-586.recode.vcf", nrows = 3000, verbose = FALSE)
american

# Convert to genind object
american = vcfR2genind(american)

# Add site IDs to genind object
american$pop = as.factor(substr(indNames(american), 1, 3))
```

[Back to Contents](#contents)

<div id="outlier"/>
#### Conduct outlier tests using OutFLANK  
Conduct _F_~ST~ differentiation-based outlier tests on `genind` object using [OutFLANK](https://github.com/whitlock/OutFLANK) using a wrapper script from the [dartR](https://cran.r-project.org/web/packages/dartR/vignettes/IntroTutorial_dartR.pdf) package.
```{r warning=FALSE, message=FALSE}
# Load packages
library(OutFLANK)
library(qvalue)
library(dartR)
```

```{r}
# Run OutFLANK using dartR wrapper script
outflnk = gl.outflank(american, qthreshold = 0.05, plot = FALSE)

# Extract OutFLANK results
outflnk.df = outflnk$outflank$results

# Remove duplicated rows for each SNP locus
rowsToRemove = seq(1, nrow(outflnk.df), by = 2)
outflnk.df = outflnk.df[-rowsToRemove, ]

# Print number of outliers (TRUE)
outflnk.df$OutlierFlag %>% summary
```

```{r}
# Extract outlier IDs
outlier_indexes = which(outflnk.df$OutlierFlag == TRUE)
outlierID = locNames(american)[outlier_indexes]
outlierID
```

```{r}
# Convert Fsts <0 to zero
outflnk.df$FST[outflnk.df$FST < 0] = 0 

# Italic labels
fstlab = expression(italic("F")[ST])
hetlab = expression(italic("H")[e])

# Plot He versus Fst
ggplot(data = outflnk.df)+
  geom_point(aes(x = He, y = FST, colour = OutlierFlag))+
  scale_colour_manual(values = c("black","red"), labels = c("Neutral SNP","Outlier SNP"))+
  ggtitle("OutFLANK outlier test")+
  xlab(hetlab)+
  ylab(fstlab)+
  theme(legend.title = element_blank(),
        plot.title = element_text(hjust = 0.5, size = 15, face = "bold")
        )

```

[Back to Contents](#contents)

<div id="export"/>
#### Export biallelic SNP genind object  
Load required packages.
```{r warning=FALSE, message=FALSE}
library(devtools)
library(miscTools)
library(stringr)
```

Export genind object in genepop format.
```{r warning=FALSE, message=FALSE, eval=FALSE}
source_url("https://raw.githubusercontent.com/Tom-Jenkins/utility_scripts/master/TJ_genind2genepop_function.R")
genind2genepop(lobster_gen_sub, file = "lobster_genotypes.gen")
```

Export genind object in STRUCTURE format. If you want to run STRUCTURE in Linux then use `unix = TRUE` which exports a Unix text file (Windows text file by default).
```{r warning=FALSE, message=FALSE, eval=FALSE}
source_url("https://raw.githubusercontent.com/Tom-Jenkins/utility_scripts/master/TJ_genind2structure_function.R")
genind2structure(lobster_gen_sub, file = "lobster_genotypes.str", pops = TRUE, markers = TRUE, unix = FALSE)
```

[Back to Contents](#contents)

<div id="resources"/>
#### Further resources  
[Conduct and visualise admixture analyses in R](https://github.com/Tom-Jenkins/admixture_pie_chart_map_tutorial)  
[Population genetics and genomics in R](https://grunwaldlab.github.io/Population_Genetics_in_R/TOC.html)  
[Detecting multilocus adaptation using redundancy analysis](https://popgen.nescent.org/2018-03-27_RDA_GEA.html)  
[Using pcadapt to detect local adaptation](https://bcm-uga.github.io/pcadapt/articles/pcadapt.html)  
[Analysis of multilocus genotypes and lineages in poppr](https://grunwaldlab.github.io/poppr/articles/mlg.html)  
[Spatial analysis of principal components analysis using adegenet](http://adegenet.r-forge.r-project.org/files/tutorial-spca.pdf)  

#### Download a PDF of this post
[Download PDF](/files/r-popgen-getting-started-pdf.pdf)

#### R session info  
```{r}
sessionInfo()
```

[Back to Contents](#contents)