---
title: "Download data from the SRA, ENA, and GenBank"
date: '2022-09-14'
diagram: yes
math: yes
output:
  blogdown::html_page:
    toc: true
    number_sections: yes
    highlight: google-light
slug: r-download-sra-ena-genbank
summary: Download data from the SRA, ENA, and GenBank via the Linux command-line.
weight: 1
image:
# Featured image placement options:
#   1 = Full column width
#   2 = Out-set
#   3 = Screen-width
  placement: 1
  caption: ""
---

```{r setup, include=FALSE}
# Remove hashes from code output
knitr::opts_chunk$set(comment = NA, collapse = TRUE)
```

```{css, echo=FALSE}
pre {
  max-height: 300px;
  overflow-y: auto;
}
```

```{css, echo=FALSE}
/* Enforce period after section number */
.header-section-number::after {
  content: ".";
}
```

<br>

<!-- START OF RMARKDOWN DOCUMENT -->
The aim of this tutorial is to demonstrate how you can download data from the Sequence Read Archive, the European Nucleotide Archive, and GenBank via the Linux command-line. 

<br>

**Prerequisites for this tutorial:**  

* A Linux computer
* 1GB storage space available on your machine*

<br>

# Download data from the SRA / ENA
The Sequence Read Archive ([SRA](https://www.ncbi.nlm.nih.gov/sra)) and the European Nucleotide Archive ([ENA](https://www.ebi.ac.uk/ena/browser/home)) are public repositories that store sequence data; at present, these databases mostly consist of short-read data generated by high-throughput sequencing. You can download raw data from both databases (usually in FASTQ format) using the general-purpose tool [fastq-dl](https://github.com/rpetit3/fastq-dl) --- for more extensive queries to each database, the NCBI (National Center for Biotechnology Information) and the EBI (European Bioinformatics Institute) provide their own tools: [SRA Toolkit](https://github.com/ncbi/sra-tools) and [enaBrowserTools](https://github.com/enasequence/enaBrowserTools).

We will download sequence data for bovine tuberculosis (*Mycobacterium bovis*) by parsing accession numbers. The accession number is an ID that is unique to a particular project (study), sequencing experiment or sequencing run. Therefore, this is what we will use when downloading sequencing data for a study or sample of interest.

Accession ID for each level:

* Project: [PRJNA804719](https://www.ebi.ac.uk/ena/browser/view/PRJNA804719)
* Experiment: [SRX14525734](https://www.ebi.ac.uk/ena/browser/view/SRX14525734?show=reads)
* Run: [SRR18391668](https://www.ebi.ac.uk/ena/browser/view/SRR18391668?show=reads)

[Back to Contents](#top)

<br>

## fastq-dl

**Installation**  
One way to install fastq-dl is to via conda from the [Bioconda](https://bioconda.github.io/) channel. By following the commands below, you will set up a conda environmental called fastq-dl which downloads fastq-dl and all of its dependencies to the fastq-dl environment.

```{bash, eval=FALSE}
conda create -n fastq-dl -c conda-forge -c bioconda fastq-dl
conda activate fastq-dl
```
> Note: installing via conda requires you to have installed [miniconda3](https://docs.conda.io/en/latest/miniconda.html#) or [miniforge3](https://github.com/conda-forge/miniforge)
>
> Tip: to deactivate a conda environment type `conda activate base`

[Back to Contents](#top)

<br>

**Download reads**  
Download raw reads for a single sequencing run to the current working directory.
```{bash, eval=FALSE}
# Download from the SRA
fastq-dl --outdir . --cpus 4 SRR18391668 SRA

# Download from the ENA
fastq-dl --outdir . --cpus 4 SRR18391668 ENA
```

```{r, eval=TRUE, echo=FALSE}
library(gt)
library(magrittr)
tibble::tribble(
  ~Argument, ~Description,
  "--outdir", "path to the directory to store output files",
  "--cpus", "number of processors to use"
  ) %>% 
  gt() %>% 
  cols_align(
    align = "left"
  )
```
> Note: for paired reads, the fastq-dl default is to download two separate gzipped FASTQ files

You can also specify a project or experiment accession ID. For example, supplying the project accession, PRJNA804719, tells the program to download all 136 sequencing runs to the current directory. Similarly, supplying the experiment accession, SRX14525734, tells the program to download all sequencing runs for that experiment. In the example used here, there is only one run in the experiment so only one sample of paired reads will be downloaded.
```{bash, eval=FALSE}
# Download all sequencing experiments for Project PRJNA804719
fastq-dl --outdir . --cpus 4 PRJNA804719 SRA # SRA
fastq-dl --outdir . --cpus 4 PRJNA804719 ENA # ENA

# Download all sequencing runs for Experiment SRX14525734
fastq-dl --outdir . --cpus 4 SRX14525734 SRA # SRA
fastq-dl --outdir . --cpus 4 SRX14525734 ENA # ENA
```
> *Note: to download all 136 sequencing runs you will need approximately 20GB storage space

[Back to Contents](#top)

<br>

# Download data from GenBank
[GenBank](https://www.ncbi.nlm.nih.gov/genbank/) is a genetic sequence database on the NCBI. It is designed to "*provide and encourage access within the scientific community to the most up-to-date and comprehensive DNA sequence information*". You can download data using the command-line tools provided by Entrez Direct ([EDirect](https://www.ncbi.nlm.nih.gov/books/NBK179288/)). 

**Installation**  
To install EDirect tools to your home directory, follow the commands below:
```{bash, eval=FALSE}
sh -c "$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)"
# Type "y" and press the Return key to add the path to your .bashrc file
# Type "N" if you want to move the edirect folder and manually add a path to your .bashrc file
```

**Examples**  
This tutorial presents examples for downloading the following data from GenBank:

* Nucleotide DNA sequences (nuclear and organelle)
* Protein sequences
* Gene information
* Genome assemblies

[Back to Contents](#top)

<br>


## Nucleotide

**Single 16S rRNA sequence**  
Download the [AF276989](https://www.ncbi.nlm.nih.gov/nuccore/AF276989) 16S rRNA sequence for _Salmonella enterica_ in FASTA format.
```{bash, eval=FALSE}
efetch -db nucleotide -id AF276989 -format fasta > AF276989_16S.fa
```

**Multiple 16S rRNA sequences**  
Download [multiple](https://www.ncbi.nlm.nih.gov/nuccore/?term=Salmonella+enterica+16S+rRNA) 16S rRNA sequences for _Salmonella enterica_ in FASTA format.
```{bash, eval=FALSE}
efetch -db nucleotide -id AF276989,NR_074799,NR_041696,NR_074910,NR_119108 -format fasta > Senterica_16S.fa
```

**Multiple 16S rRNA sequences --- text file containing IDs**  
Same as above but instead submit a text file containing the IDs to download.
```
# Text file called nucleotide_ids.txt
AF276989
NR_074799
NR_041696
NR_074910
NR_119108
```
```{bash, eval=FALSE}
efetch -db nucleotide -input nucleotide_ids.txt -format fasta > Senterica_16S.fa
```

**Mitogenome**  
Download the complete mitochondrion genome ([MH747083](https://www.ncbi.nlm.nih.gov/nuccore/MH747083)) for _Homarus gammarus_ in FASTA format.
```{bash, eval=FALSE}
efetch -db nucleotide -id MH747083 -format fasta > Hgammarus_mitogenome.fa
```

**Plastome**  
Download the complete plastid genome ([MH281627](https://www.ncbi.nlm.nih.gov/nuccore/MH281627)) for _Lithothamnion sp._ in FASTA format.
```{bash, eval=FALSE}
efetch -db nucleotide -id MH281627 -format fasta > LithothamnionSp_plastome.fa
```

[Back to Contents](#top)

<br>

## Protein

**Single protein sequence**  
Download the [NP_776541](https://www.ncbi.nlm.nih.gov/protein/NP_776541.1) protein sequence for myogenic factor 5 (*Bos taurus*) in FASTA format.
```{bash, eval=FALSE}
efetch -db protein -id NP_776541 -format fasta > Btaurus_MYF5.fa
```

**Multiple protein sequences**  
Download [multiple](https://www.ncbi.nlm.nih.gov/protein/?term=Homo+sapiens) protein sequences for humans in FASTA format.
```{bash, eval=FALSE}
efetch -db protein -id NP_001186746,AAB51177,AAB51171,AAF88103,AAF88103 -format fasta > Human_proteins.fa
```

**Multiple 16S rRNA sequences --- text file containing IDs**  
Same as above but instead submit a text file containing the IDs to download.
```
# Text file called protein_ids.txt
NP_001186746
AAB51177
AAB51171
AAF88103
AAF88103
```
```{bash, eval=FALSE}
efetch -db protein -input protein_ids.txt -format fasta > Human_proteins.fa
```

[Back to Contents](#top)

<br>

## Gene

**Single gene**  
Download information for the Mbis1 gene  ([492906](https://www.ncbi.nlm.nih.gov/gene/492906)) from mouse (*Mus musculus*) in tabular format.
```{bash, eval=FALSE}
# Method 1: use gene ID
efetch -db gene -id 492906 -format tabular > Mouse_Mbis1.tsv

# Method 2: use a descriptive query
esearch -db gene -query "Mus musculus Mbis1" | efetch -format tabular > Mouse_Mbis1.tsv
```

**Multiple genes**  
Download information for [multiple](https://www.ncbi.nlm.nih.gov/gene/?term=Danio+rerio) genes from zebrafish (*Danio rerio*) in tabular format.
```{bash, eval=FALSE}
efetch -db gene -id 30592,30513,30425,30590,30269 -format tabular > Zebrafish_genes.tsv
```

**Multiple genes --- text file containing IDs**  
Same as above but instead submit a text file containing the IDs to download.
```
# Text file called gene_ids.txt
30592
30513
30425
30590
30269
```
```{bash, eval=FALSE}
efetch -db gene -input gene_ids.txt -format tabular > Zebrafish_genes.tsv
```

[Back to Contents](#top)

<br>

## Assembly

**_Mycobacterium bovis_**  
Download _Mycobacterium bovis_ complete genome assembly ([GCA_005156105.1](https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_005156105.1/)) in gzipped FASTA format.
```{bash, eval=FALSE}
wget $(esearch -db assembly -query "GCA_005156105.1" | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F"/" '{print $0"/"$NF"_genomic.fna.gz"}')
```

**_Homarus americanus_**  
Download _Homarus americanus_ scaffold genome assembly ([GCA_018991925.1](https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_018991925.1/)) in gzipped FASTA format.
```{bash, eval=FALSE}
wget $(esearch -db assembly -query "GCA_018991925.1" | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F"/" '{print $0"/"$NF"_genomic.fna.gz"}')
```

**_Chondrus crispus_**  
Download _Chondrus crispus_ scaffold genome assembly ([GCA_000350225.2](https://www.ncbi.nlm.nih.gov/data-hub/genome/GCA_000350225.2/)) in gzipped FASTA format.
```{bash, eval=FALSE}
wget $(esearch -db assembly -query "GCA_000350225.2" | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F"/" '{print $0"/"$NF"_genomic.fna.gz"}')
```

[Back to Contents](#top)

<br>

> Further information for downloading data via EDirect can be found on the following pages:
>
>> https://www.ncbi.nlm.nih.gov/books/NBK179288/  
>> https://github.com/NCBI-Hackathons/EDirectCookbook

<br>

# R session info
This tutorial was created using R Markdown and Knitr.
```{r, echo=FALSE, eval=TRUE}
library(rmarkdown)
library(knitr)
library(magrittr)
library(gt)
sessionInfo()
```

[Back to Contents](#top)
