---
title: "Download data from the SRA, ENA, and GenBank"
date: '2022-09-14'
diagram: yes
math: yes
output:
  blogdown::html_page:
    toc: true
    number_sections: yes
slug: r-download-sra-ena-genbank
summary: Download data from the SRA, ENA, and GenBank via the Linux command-line.
weight: 1
image:
# Featured image placement options:
#   1 = Full column width
#   2 = Out-set
#   3 = Screen-width
  placement: 1
  caption: ""
---


<div id="TOC">
<ul>
<li><a href="#download-data-from-the-sra-ena"><span class="toc-section-number">1</span> Download data from the SRA / ENA</a><ul>
<li><a href="#fastq-dl"><span class="toc-section-number">1.1</span> fastq-dl</a></li>
</ul></li>
<li><a href="#download-data-from-genbank"><span class="toc-section-number">2</span> Download data from GenBank</a><ul>
<li><a href="#nucleotide"><span class="toc-section-number">2.1</span> Nucleotide</a></li>
<li><a href="#protein"><span class="toc-section-number">2.2</span> Protein</a></li>
<li><a href="#gene"><span class="toc-section-number">2.3</span> Gene</a></li>
<li><a href="#assembly"><span class="toc-section-number">2.4</span> Assembly</a></li>
</ul></li>
<li><a href="#r-session-info"><span class="toc-section-number">3</span> R session info</a></li>
</ul>
</div>

<style type="text/css">
pre {
  max-height: 300px;
  overflow-y: auto;
}
</style>
<style type="text/css">
/* Enforce period after section number */
.header-section-number::after {
  content: ".";
}
</style>
<!-- START OF RMARKDOWN DOCUMENT -->
<p>The aim of this tutorial is to demonstrate how you can download data from the Sequence Read Archive, the European Nucleotide Archive, and GenBank via the Linux command-line.</p>
<p><strong>Prerequisites for this tutorial:</strong></p>
<ul>
<li>A Linux computer</li>
<li>1GB storage space available on your machine*</li>
</ul>
<p><br></p>
<div id="download-data-from-the-sra-ena" class="section level1">
<h1><span class="header-section-number">1</span> Download data from the SRA / ENA</h1>
<p>The Sequence Read Archive (<a href="https://www.ncbi.nlm.nih.gov/sra">SRA</a>) and the European Nucleotide Archive (<a href="https://www.ebi.ac.uk/ena/browser/home">ENA</a>) are public repositories that store sequence data; at present, these databases mostly consist of short-read data generated by high-throughput sequencing. You can download raw data from both databases (usually in FASTQ format) using the general-purpose tool <a href="https://github.com/rpetit3/fastq-dl">fastq-dl</a> — for more extensive queries to each database, the NCBI (National Center for Biotechnology Information) and the EBI (European Bioinformatics Institute) provide their own tools: <a href="https://github.com/ncbi/sra-tools">SRA Toolkit</a> and <a href="https://github.com/enasequence/enaBrowserTools">enaBrowserTools</a>.</p>
<p>We will download sequence data for bovine tuberculosis (<em>Mycobacterium bovis</em>) by parsing accession numbers. The accession number is an ID that is unique to a particular project (study), sequencing experiment or sequencing run. Therefore, this is what we will use when downloading sequencing data for a study or sample of interest.</p>
<p>Accession ID for each level:</p>
<ul>
<li>Project: <a href="https://www.ebi.ac.uk/ena/browser/view/PRJNA804719">PRJNA804719</a></li>
<li>Experiment: <a href="https://www.ebi.ac.uk/ena/browser/view/SRX14525734?show=reads">SRX14525734</a></li>
<li>Run: <a href="https://www.ebi.ac.uk/ena/browser/view/SRR18391668?show=reads">SRR18391668</a></li>
</ul>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
<div id="fastq-dl" class="section level2">
<h2><span class="header-section-number">1.1</span> fastq-dl</h2>
<p><strong>Installation</strong><br />
One way to install fastq-dl is to via conda from the <a href="https://bioconda.github.io/">Bioconda</a> channel. By following the commands below, you will set up a conda environmental called fastq-dl which downloads fastq-dl and all of its dependencies to the fastq-dl environment.</p>
<pre class="bash"><code>conda create -n fastq-dl -c conda-forge -c bioconda fastq-dl
conda activate fastq-dl</code></pre>
<blockquote>
<p>Note: installing via conda requires you to have installed <a href="https://docs.conda.io/en/latest/miniconda.html#">miniconda3</a> or <a href="https://github.com/conda-forge/miniforge">miniforge3</a></p>
<p>Tip: to deactivate a conda environment type <code>conda activate base</code></p>
</blockquote>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
<p><strong>Download reads</strong><br />
Download raw reads for a single sequencing run to the current working directory.</p>
<pre class="bash"><code># Download from the SRA
fastq-dl --outdir . --cpus 4 SRR18391668 SRA

# Download from the ENA
fastq-dl --outdir . --cpus 4 SRR18391668 ENA</code></pre>
<div id="fzsocsvphm" style="overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>html {
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Helvetica Neue', 'Fira Sans', 'Droid Sans', Arial, sans-serif;
}

#fzsocsvphm .gt_table {
  display: table;
  border-collapse: collapse;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#fzsocsvphm .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#fzsocsvphm .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#fzsocsvphm .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 0;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#fzsocsvphm .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#fzsocsvphm .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#fzsocsvphm .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#fzsocsvphm .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#fzsocsvphm .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#fzsocsvphm .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#fzsocsvphm .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#fzsocsvphm .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
}

#fzsocsvphm .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#fzsocsvphm .gt_from_md > :first-child {
  margin-top: 0;
}

#fzsocsvphm .gt_from_md > :last-child {
  margin-bottom: 0;
}

#fzsocsvphm .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#fzsocsvphm .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#fzsocsvphm .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#fzsocsvphm .gt_row_group_first td {
  border-top-width: 2px;
}

#fzsocsvphm .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#fzsocsvphm .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#fzsocsvphm .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#fzsocsvphm .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#fzsocsvphm .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#fzsocsvphm .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#fzsocsvphm .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#fzsocsvphm .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#fzsocsvphm .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#fzsocsvphm .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-left: 4px;
  padding-right: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#fzsocsvphm .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#fzsocsvphm .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#fzsocsvphm .gt_left {
  text-align: left;
}

#fzsocsvphm .gt_center {
  text-align: center;
}

#fzsocsvphm .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#fzsocsvphm .gt_font_normal {
  font-weight: normal;
}

#fzsocsvphm .gt_font_bold {
  font-weight: bold;
}

#fzsocsvphm .gt_font_italic {
  font-style: italic;
}

#fzsocsvphm .gt_super {
  font-size: 65%;
}

#fzsocsvphm .gt_footnote_marks {
  font-style: italic;
  font-weight: normal;
  font-size: 75%;
  vertical-align: 0.4em;
}

#fzsocsvphm .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#fzsocsvphm .gt_indent_1 {
  text-indent: 5px;
}

#fzsocsvphm .gt_indent_2 {
  text-indent: 10px;
}

#fzsocsvphm .gt_indent_3 {
  text-indent: 15px;
}

#fzsocsvphm .gt_indent_4 {
  text-indent: 20px;
}

#fzsocsvphm .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table">
  
  <thead class="gt_col_headings">
    <tr>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">Argument</th>
      <th class="gt_col_heading gt_columns_bottom_border gt_left" rowspan="1" colspan="1" scope="col">Description</th>
    </tr>
  </thead>
  <tbody class="gt_table_body">
    <tr><td class="gt_row gt_left">--outdir</td>
<td class="gt_row gt_left">path to the directory to store output files</td></tr>
    <tr><td class="gt_row gt_left">--cpus</td>
<td class="gt_row gt_left">number of processors to use</td></tr>
  </tbody>
  
  
</table>
</div>
<blockquote>
<p>Note: for paired reads, the fastq-dl default is to download two separate gzipped FASTQ files</p>
</blockquote>
<p>You can also specify a project or experiment accession ID. For example, supplying the project accession, PRJNA804719, tells the program to download all 136 sequencing runs to the current directory. Similarly, supplying the experiment accession, SRX14525734, tells the program to download all sequencing runs for that experiment. In the example used here, there is only one run in the experiment so only one sample of paired reads will be downloaded.</p>
<pre class="bash"><code># Download all sequencing experiments for Project PRJNA804719
fastq-dl --outdir . --cpus 4 PRJNA804719 SRA # SRA
fastq-dl --outdir . --cpus 4 PRJNA804719 ENA # ENA

# Download all sequencing runs for Experiment SRX14525734
fastq-dl --outdir . --cpus 4 SRX14525734 SRA # SRA
fastq-dl --outdir . --cpus 4 SRX14525734 ENA # ENA</code></pre>
<blockquote>
<p>*Note: to download all 136 sequencing runs you will need approximately 20GB storage space</p>
</blockquote>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
</div>
</div>
<div id="download-data-from-genbank" class="section level1">
<h1><span class="header-section-number">2</span> Download data from GenBank</h1>
<p><a href="https://www.ncbi.nlm.nih.gov/genbank/">GenBank</a> is a genetic sequence database on the NCBI. It is designed to “<em>provide and encourage access within the scientific community to the most up-to-date and comprehensive DNA sequence information</em>”. You can download data using the command-line tools provided by Entrez Direct (<a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/">EDirect</a>).</p>
<p><strong>Installation</strong><br />
To install EDirect tools to your home directory, follow the commands below:</p>
<pre class="bash"><code>sh -c &quot;$(curl -fsSL ftp://ftp.ncbi.nlm.nih.gov/entrez/entrezdirect/install-edirect.sh)&quot;
# Type &quot;y&quot; and press the Return key to add the path to your .bashrc file
# Type &quot;N&quot; if you want to move the edirect folder and manually add a path to your .bashrc file</code></pre>
<p><strong>Examples</strong><br />
This tutorial presents examples for downloading the following data from GenBank:</p>
<ul>
<li>Nucleotide DNA sequences (nuclear and organelle)</li>
<li>Protein sequences</li>
<li>Gene information</li>
<li>Genome assemblies</li>
</ul>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
<div id="nucleotide" class="section level2">
<h2><span class="header-section-number">2.1</span> Nucleotide</h2>
<p><strong>Single 16S rRNA sequence</strong><br />
Download the <a href="https://www.ncbi.nlm.nih.gov/nuccore/AF276989">AF276989</a> 16S rRNA sequence for <em>Salmonella enterica</em> in FASTA format.</p>
<pre class="bash"><code>efetch -db nucleotide -id AF276989 -format fasta &gt; AF276989_16S.fa</code></pre>
<p><strong>Multiple 16S rRNA sequences</strong><br />
Download <a href="https://www.ncbi.nlm.nih.gov/nuccore/?term=Salmonella+enterica+16S+rRNA">multiple</a> 16S rRNA sequences for <em>Salmonella enterica</em> in FASTA format.</p>
<pre class="bash"><code>efetch -db nucleotide -id AF276989,NR_074799,NR_041696,NR_074910,NR_119108 -format fasta &gt; Senterica_16S.fa</code></pre>
<p><strong>Multiple 16S rRNA sequences — text file containing IDs</strong><br />
Same as above but instead submit a text file containing the IDs to download.</p>
<pre><code># Text file called nucleotide_ids.txt
AF276989
NR_074799
NR_041696
NR_074910
NR_119108</code></pre>
<pre class="bash"><code>efetch -db nucleotide -input nucleotide_ids.txt -format fasta &gt; Senterica_16S.fa</code></pre>
<p><strong>Mitogenome</strong><br />
Download the complete mitochondrion genome (<a href="https://www.ncbi.nlm.nih.gov/nuccore/MH747083">MH747083</a>) for <em>Homarus gammarus</em> in FASTA format.</p>
<pre class="bash"><code>efetch -db nucleotide -id MH747083 -format fasta &gt; Hgammarus_mitogenome.fa</code></pre>
<p><strong>Plastome</strong><br />
Download the complete plastid genome (<a href="https://www.ncbi.nlm.nih.gov/nuccore/MH281627">MH281627</a>) for <em>Lithothamnion sp.</em> in FASTA format.</p>
<pre class="bash"><code>efetch -db nucleotide -id MH281627 -format fasta &gt; LithothamnionSp_plastome.fa</code></pre>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
</div>
<div id="protein" class="section level2">
<h2><span class="header-section-number">2.2</span> Protein</h2>
<p><strong>Single protein sequence</strong><br />
Download the <a href="https://www.ncbi.nlm.nih.gov/protein/NP_776541.1">NP_776541</a> protein sequence for myogenic factor 5 (<em>Bos taurus</em>) in FASTA format.</p>
<pre class="bash"><code>efetch -db protein -id NP_776541 -format fasta &gt; Btaurus_MYF5.fa</code></pre>
<p><strong>Multiple protein sequences</strong><br />
Download <a href="https://www.ncbi.nlm.nih.gov/protein/?term=Homo+sapiens">multiple</a> protein sequences for humans in FASTA format.</p>
<pre class="bash"><code>efetch -db protein -id NP_001186746,AAB51177,AAB51171,AAF88103,AAF88103 -format fasta &gt; Human_proteins.fa</code></pre>
<p><strong>Multiple 16S rRNA sequences — text file containing IDs</strong><br />
Same as above but instead submit a text file containing the IDs to download.</p>
<pre><code># Text file called protein_ids.txt
NP_001186746
AAB51177
AAB51171
AAF88103
AAF88103</code></pre>
<pre class="bash"><code>efetch -db protein -input protein_ids.txt -format fasta &gt; Human_proteins.fa</code></pre>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
</div>
<div id="gene" class="section level2">
<h2><span class="header-section-number">2.3</span> Gene</h2>
<p><strong>Single gene</strong><br />
Download information for the Mbis1 gene (<a href="https://www.ncbi.nlm.nih.gov/gene/492906">492906</a>) from mouse (<em>Mus musculus</em>) in tabular format.</p>
<pre class="bash"><code># Method 1: use gene ID
efetch -db gene -id 492906 -format tabular &gt; Mouse_Mbis1.tsv

# Method 2: use a descriptive query
esearch -db gene -query &quot;Mus musculus Mbis1&quot; | efetch -format tabular &gt; Mouse_Mbis1.tsv</code></pre>
<p><strong>Multiple genes</strong><br />
Download information for <a href="https://www.ncbi.nlm.nih.gov/gene/?term=Danio+rerio">multiple</a> genes from zebrafish (<em>Danio rerio</em>) in tabular format.</p>
<pre class="bash"><code>efetch -db gene -id 30592,30513,30425,30590,30269 -format tabular &gt; Zebrafish_genes.tsv</code></pre>
<p><strong>Multiple genes — text file containing IDs</strong><br />
Same as above but instead submit a text file containing the IDs to download.</p>
<pre><code># Text file called gene_ids.txt
30592
30513
30425
30590
30269</code></pre>
<pre class="bash"><code>efetch -db gene -input gene_ids.txt -format tabular &gt; Zebrafish_genes.tsv</code></pre>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
</div>
<div id="assembly" class="section level2">
<h2><span class="header-section-number">2.4</span> Assembly</h2>
<p><strong><em>Mycobacterium bovis</em></strong><br />
Download <em>Mycobacterium bovis</em> complete genome assembly (<a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_005156105.1/">GCA_005156105.1</a>) in gzipped FASTA format.</p>
<pre class="bash"><code>wget $(esearch -db assembly -query &quot;GCA_005156105.1&quot; | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F&quot;/&quot; &#39;{print $0&quot;/&quot;$NF&quot;_genomic.fna.gz&quot;}&#39;)</code></pre>
<p><strong><em>Homarus americanus</em></strong><br />
Download <em>Homarus americanus</em> scaffold genome assembly (<a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_018991925.1/">GCA_018991925.1</a>) in gzipped FASTA format.</p>
<pre class="bash"><code>wget $(esearch -db assembly -query &quot;GCA_018991925.1&quot; | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F&quot;/&quot; &#39;{print $0&quot;/&quot;$NF&quot;_genomic.fna.gz&quot;}&#39;)</code></pre>
<p><strong><em>Chondrus crispus</em></strong><br />
Download <em>Chondrus crispus</em> scaffold genome assembly (<a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/GCA_000350225.2/">GCA_000350225.2</a>) in gzipped FASTA format.</p>
<pre class="bash"><code>wget $(esearch -db assembly -query &quot;GCA_000350225.2&quot; | esummary | xtract -pattern DocumentSummary -element FtpPath_RefSeq | awk -F&quot;/&quot; &#39;{print $0&quot;/&quot;$NF&quot;_genomic.fna.gz&quot;}&#39;)</code></pre>
<p><a href="#top">Back to Contents</a></p>
<p><br></p>
<blockquote>
<p>Further information for downloading data via EDirect can be found on the following pages:</p>
<blockquote>
<p><a href="https://www.ncbi.nlm.nih.gov/books/NBK179288/" class="uri">https://www.ncbi.nlm.nih.gov/books/NBK179288/</a><br />
<a href="https://github.com/NCBI-Hackathons/EDirectCookbook" class="uri">https://github.com/NCBI-Hackathons/EDirectCookbook</a></p>
</blockquote>
</blockquote>
<p><br></p>
</div>
</div>
<div id="r-session-info" class="section level1">
<h1><span class="header-section-number">3</span> R session info</h1>
<p>This tutorial was created using R Markdown and Knitr.</p>
<pre><code>R version 4.2.1 (2022-06-23 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 10 x64 (build 18363)

Matrix products: default

locale:
[1] LC_COLLATE=English_United Kingdom.utf8 
[2] LC_CTYPE=English_United Kingdom.utf8   
[3] LC_MONETARY=English_United Kingdom.utf8
[4] LC_NUMERIC=C                           
[5] LC_TIME=English_United Kingdom.utf8    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] knitr_1.39     rmarkdown_2.15 magrittr_2.0.3 gt_0.7.0      

loaded via a namespace (and not attached):
 [1] bslib_0.4.0      compiler_4.2.1   pillar_1.8.1     jquerylib_0.1.4 
 [5] tools_4.2.1      digest_0.6.29    gtable_0.3.0     jsonlite_1.8.0  
 [9] evaluate_0.16    lifecycle_1.0.1  tibble_3.1.8     pkgconfig_2.0.3 
[13] rlang_1.0.4      cli_3.3.0        DBI_1.1.3        rstudioapi_0.14 
[17] yaml_2.3.5       blogdown_1.11    xfun_0.32        fastmap_1.1.0   
[21] dplyr_1.0.9      stringr_1.4.1    generics_0.1.3   vctrs_0.4.1     
[25] sass_0.4.2       grid_4.2.1       tidyselect_1.1.2 glue_1.6.2      
[29] R6_2.5.1         fansi_1.0.3      bookdown_0.28    purrr_0.3.4     
[33] ggplot2_3.3.6    ellipsis_0.3.2   scales_1.2.1     htmltools_0.5.3 
[37] assertthat_0.2.1 colorspace_2.0-3 utf8_1.2.2       stringi_1.7.8   
[41] munsell_0.5.0    cachem_1.0.6    </code></pre>
<p><a href="#top">Back to Contents</a></p>
</div>
